{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3b5810ff-72a9-4030-b9f3-081035abbf55",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "fd56d337-c379-46b8-a6d9-81e25b0b793c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nf_files(folder):\n",
    "    nf_files = []\n",
    "    for entry in os.listdir(folder):\n",
    "        if os.path.isfile(os.path.join(folder, entry)) and entry.endswith('.nf'):\n",
    "            nf_files.append(os.path.join(folder, entry))\n",
    "        if os.path.isdir(os.path.join(folder, entry)):\n",
    "            new_files = get_nf_files(os.path.join(folder, entry))\n",
    "            for file in new_files:\n",
    "                nf_files.append(file)\n",
    "    return nf_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "d8e11ecc-dc10-4f86-9b33-ebf3359b4027",
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open('BALROG-MON\\\\modules\\\\BUSCO.nf') as file:\n",
    "#    text = file.read().split('\\n')\n",
    "#    print(text[2].split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "998d87aa-52b8-439a-9d37-5f6e51ab9fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_workflows(pipeline_dir, workflow_dir):\n",
    "    nf_workflows = get_nf_files(pipeline+ '/' + workflow_dir)\n",
    "    print(pipeline_dir + ' contains these workflows: \\n')\n",
    "    for wf in nf_workflows:\n",
    "        print(wf)\n",
    "    return nf_workflows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "f7a5f94e-62db-4a32-ab1a-d40552c03cc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_used_files(pipeline_dir, workflows):\n",
    "    # Creating list to store subworkflows and modules that are used\n",
    "    used_modules = []\n",
    "    used_subworkflows = []\n",
    "    \n",
    "    for workflow in workflows:\n",
    "        print(\"Subworkflows in: \" + workflow.split('\\\\')[-1])\n",
    "        with open(workflow) as file:\n",
    "            text = file.read().split('\\n')\n",
    "            for line in text:\n",
    "                if line.split(' ')[0] == \"include\":\n",
    "                    print(line.split(' ')[-1])\n",
    "                    subworkflow = line.split(' ')[-1][4:]\n",
    "                    used_subworkflows.append(subworkflow.split('/')[-1][:-1])\n",
    "    \n",
    "                    print('modules used in: ' + subworkflow)\n",
    "                    with open(workflow.split('/')[0] + '/' + subworkflow[:-1]) as file:\n",
    "                        sw_text = file.read().split('\\n')\n",
    "                        for line in sw_text:\n",
    "                            if line.split(' ')[0] == \"include\":\n",
    "                                print(line.split(' ')[-1])\n",
    "                                used_modules.append(line.split(' ')[-1].split('/')[-1])\n",
    "                                \n",
    "    return [used_subworkflows, used_modules]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "5e86fd33-0755-4d2c-9c5e-3334841f632b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unused_files(pipeline_dir, subworkflow_dir, module_dir, used_sub, used_mod):\n",
    "## Creating list of subworkflows to move to unused\n",
    "    unused_subworkflows = []\n",
    "    unused_modules = []\n",
    "    \n",
    "    for sub in os.listdir(pipeline_dir + '/' + subworkflow_dir):\n",
    "        if sub not in used_sub:\n",
    "            print(sub + ' is not used in this pipeline and will be scheduled for archival')\n",
    "            unused_subworkflows.append(sub)\n",
    "    \n",
    "    print('---------------------------------------------------------------')\n",
    "    \n",
    "    for mod in os.listdir(pipeline_dir + '/' + module_dir):\n",
    "        if mod not in used_mod:\n",
    "            print(mod + ' is not used in this pipeline and will be scheduled for archival')\n",
    "            unused_modules.append(mod)\n",
    "\n",
    "    return [unused_subworkflows, unused_modules]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "30bcb93f-2be0-49cc-8e6a-5a157ee872cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_archive_structure(pipeline_dir): \n",
    "    # Define the directory paths\n",
    "    main_dir = pipeline_dir + '/Archive_NFParser'\n",
    "    sub_dirs = [\"modules\", \"subworkflows\"]\n",
    "    \n",
    "    # Create the main directory\n",
    "    os.makedirs(main_dir, exist_ok=True)\n",
    "    \n",
    "    # Create subdirectories inside the main directory\n",
    "    for sub_dir in sub_dirs:\n",
    "        os.makedirs(os.path.join(main_dir, sub_dir), exist_ok=True)\n",
    "    \n",
    "    print(\"Directory structure created (or already exists).\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "df9ab800-4c4a-4d0f-94f0-faa4b142747f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def move_file_with_suffix(src_file, dest_dir):\n",
    "    # Get the base name and extension of the file\n",
    "    base_name, ext = os.path.splitext(os.path.basename(src_file))\n",
    "    dest_file = os.path.join(dest_dir, os.path.basename(src_file))\n",
    "    \n",
    "    # Check if the file already exists\n",
    "    counter = 1\n",
    "    while os.path.exists(dest_file):\n",
    "        # Modify the file name with a suffix (e.g., -1, -2, etc.)\n",
    "        dest_file = os.path.join(dest_dir, f\"{base_name}-{counter}{ext}\")\n",
    "        counter += 1\n",
    "\n",
    "    # Move the file\n",
    "    shutil.move(src_file, dest_file)\n",
    "    print(f\"Moved: {src_file} to {dest_file}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "716d3d2e-045a-4b49-a615-23ac352bd1c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def archive_files(pipeline_dir, source_dir, file_list, archive_folder):\n",
    "    if len(file_list) > 0:\n",
    "        for file in file_list:\n",
    "            move_file_with_suffix(pipeline_dir + '/' + source_dir + '/' + file, pipeline_dir + '/Archive_NFParser/' + archive_folder)\n",
    "    else:\n",
    "        print('There were no unused files in ' + source_dir + ' so no files were moved')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "97efdf8a-7c9f-418a-9e0e-83754e874dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Args\n",
    "pipeline_dir = 'BALROG-MON-TEST'\n",
    "workflow_dir = 'workflows'\n",
    "subworkflow_dir = 'subworkflows'\n",
    "module_dir = 'modules'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "715eaf8d-8093-4577-80f2-ceb73dba7581",
   "metadata": {},
   "outputs": [],
   "source": [
    "def whole_process(pipeline_dir, workflow_dir, subworkflow_dir, module_dir):\n",
    "    workflows = get_workflows(pipeline_dir, workflow_dir)\n",
    "    used_files = get_used_files(pipeline_dir, workflows)\n",
    "    unused_files = get_unused_files(pipeline_dir, subworkflow_dir, module_dir, used_files[0], used_files[1])\n",
    "    create_archive_structure(pipeline_dir)\n",
    "    archive_files(pipeline_dir, subworkflow_dir, unused_files[0], subworkflow_dir)\n",
    "    archive_files(pipeline_dir, module_dir, unused_files[1], module_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "92a6e2d2-ca67-4483-a89e-751477f847ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BALROG-MON-TEST contains these workflows: \n",
      "\n",
      "BALROG-MON/workflows\\ONT_METAGENOMIC.nf\n",
      "BALROG-MON/workflows\\ONT_METAGENOMIC_RNA.nf\n",
      "BALROG-MON/workflows\\SHORT_READ_ISOLATE.nf\n",
      "BALROG-MON/workflows\\SHORT_READ_METAGENOMIC.nf\n",
      "BALROG-MON/workflows\\SHORT_READ_METAGENOMIC_RNA.nf\n",
      "Subworkflows in: ONT_METAGENOMIC.nf\n",
      "'../subworkflows/READ_QC_ONT.nf'\n",
      "modules used in: subworkflows/READ_QC_ONT.nf'\n",
      "'../modules/FASTQC_ONT.nf'\n",
      "'../modules/FASTQC_ONT.nf'\n",
      "'../modules/FASTQC_ONT.nf'\n",
      "'../modules/PORECHOP.nf'\n",
      "'../modules/CHOPPER.nf'\n",
      "'../modules/RENAME_READS.nf'\n",
      "'../subworkflows/PLASMID_PREDICTION.nf'\n",
      "modules used in: subworkflows/PLASMID_PREDICTION.nf'\n",
      "'../modules/PLASMER.nf'\n",
      "'../modules/PLASMER_SORT.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../modules/VIRALVERIFY.nf'\n",
      "'../modules/PLASMER_DB.nf'\n",
      "'../modules/PFAM_DB.nf'\n",
      "'../modules/VIRSORTER2.nf'\n",
      "'../modules/VIRSORTER2_DB.nf'\n",
      "'../modules/MEF.nf'\n",
      "'../subworkflows/METAGENOMIC_COMMUNITY_ANALYSIS_ONT.nf'\n",
      "modules used in: subworkflows/METAGENOMIC_COMMUNITY_ANALYSIS_ONT.nf'\n",
      "'../modules/GTDBTK_DB.nf'\n",
      "'../modules/SYLPH_SKETCH_GTDB.nf'\n",
      "'../modules/SYLPH_PROFILE_SE.nf'\n",
      "'../modules/KRAKEN2_DB_PLUSPF.nf'\n",
      "'../modules/KRAKEN2_PLUSPF_SE.nf'\n",
      "'../modules/KRAKEN_BRACKEN.nf'\n",
      "'../modules/KRAKEN_ALPHADIV.nf'\n",
      "'../modules/KRAKEN_BETADIV.nf'\n",
      "'../modules/KRAKEN_REPORT2MPA.nf'\n",
      "'../modules/KRAKEN_REPORT2KRONA.nf'\n",
      "'../subworkflows/HOST_REMOVAL_ONT.nf'\n",
      "modules used in: subworkflows/HOST_REMOVAL_ONT.nf'\n",
      "'../modules/MINIMAP2_ONT.nf'\n",
      "'../modules/SAMTOOLS_EXTRACT_UNMAPPED_ONT.nf'\n",
      "'../modules/SAMTOOLS_UNMAPPED_ONT.nf'\n",
      "'../modules/FASTQC_ONT.nf'\n",
      "'../modules/SAMTOOLS_STATS.nf'\n",
      "'../modules/SAMTOOLS_STATS.nf'\n",
      "'../modules/SAMTOOLS_READNAMES.nf'\n",
      "'../modules/GUNZIP_ONT.nf'\n",
      "'../modules/SEQTK_SUBSEQ_ONT.nf'\n",
      "'../modules/CONCAT_FA.nf'\n",
      "'../subworkflows/ONT_ASSEMBLY.nf'\n",
      "modules used in: subworkflows/ONT_ASSEMBLY.nf'\n",
      "'../modules/FLYE_META.nf'\n",
      "'../modules/SEQTK_FQ2FA.nf'\n",
      "'../subworkflows/MULTI_AMR.nf'\n",
      "modules used in: subworkflows/MULTI_AMR.nf'\n",
      "'../modules/RESFINDER_DB.nf'\n",
      "'../modules/AMRFINDER_DB.nf'\n",
      "'../modules/CARD_DB.nf'\n",
      "'../modules/AMRFINDER.nf'\n",
      "'../modules/CARD_CONTIG.nf'\n",
      "'../modules/RESFINDER.nf'\n",
      "'../modules/AMRFINDER_PARSE.nf'\n",
      "'../modules/CARD_PARSE.nf'\n",
      "'../modules/RESFINDER_PARSE.nf'\n",
      "'../subworkflows/METAGENOMIC_SEQUENCE_IDENTIFICATION.nf'\n",
      "modules used in: subworkflows/METAGENOMIC_SEQUENCE_IDENTIFICATION.nf'\n",
      "'../modules/KRAKEN2_PLUSPF_FA.nf'\n",
      "'../modules/KRAKEN2_DB_PLUSPF.nf'\n",
      "Subworkflows in: ONT_METAGENOMIC_RNA.nf\n",
      "'../subworkflows/READ_QC_ONT.nf'\n",
      "modules used in: subworkflows/READ_QC_ONT.nf'\n",
      "'../modules/FASTQC_ONT.nf'\n",
      "'../modules/FASTQC_ONT.nf'\n",
      "'../modules/FASTQC_ONT.nf'\n",
      "'../modules/PORECHOP.nf'\n",
      "'../modules/CHOPPER.nf'\n",
      "'../modules/RENAME_READS.nf'\n",
      "'../subworkflows/METAGENOMIC_COMMUNITY_ANALYSIS_ONT.nf'\n",
      "modules used in: subworkflows/METAGENOMIC_COMMUNITY_ANALYSIS_ONT.nf'\n",
      "'../modules/GTDBTK_DB.nf'\n",
      "'../modules/SYLPH_SKETCH_GTDB.nf'\n",
      "'../modules/SYLPH_PROFILE_SE.nf'\n",
      "'../modules/KRAKEN2_DB_PLUSPF.nf'\n",
      "'../modules/KRAKEN2_PLUSPF_SE.nf'\n",
      "'../modules/KRAKEN_BRACKEN.nf'\n",
      "'../modules/KRAKEN_ALPHADIV.nf'\n",
      "'../modules/KRAKEN_BETADIV.nf'\n",
      "'../modules/KRAKEN_REPORT2MPA.nf'\n",
      "'../modules/KRAKEN_REPORT2KRONA.nf'\n",
      "'../subworkflows/MULTI_AMR.nf'\n",
      "modules used in: subworkflows/MULTI_AMR.nf'\n",
      "'../modules/RESFINDER_DB.nf'\n",
      "'../modules/AMRFINDER_DB.nf'\n",
      "'../modules/CARD_DB.nf'\n",
      "'../modules/AMRFINDER.nf'\n",
      "'../modules/CARD_CONTIG.nf'\n",
      "'../modules/RESFINDER.nf'\n",
      "'../modules/AMRFINDER_PARSE.nf'\n",
      "'../modules/CARD_PARSE.nf'\n",
      "'../modules/RESFINDER_PARSE.nf'\n",
      "'../subworkflows/ONT_ASSEMBLY.nf'\n",
      "modules used in: subworkflows/ONT_ASSEMBLY.nf'\n",
      "'../modules/FLYE_META.nf'\n",
      "'../modules/SEQTK_FQ2FA.nf'\n",
      "Subworkflows in: SHORT_READ_ISOLATE.nf\n",
      "'../subworkflows/READ_QC_SR.nf'\n",
      "modules used in: subworkflows/READ_QC_SR.nf'\n",
      "'../modules/FASTQC.nf'\n",
      "'../modules/FASTQC.nf'\n",
      "'../modules/FASTP.nf'\n",
      "'../subworkflows/SHORT_READ_ISOLATE_ASSEMBLY.nf'\n",
      "modules used in: subworkflows/SHORT_READ_ISOLATE_ASSEMBLY.nf'\n",
      "'../modules/SPADES_GENOME.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../modules/BUSCO.nf'\n",
      "'../modules/BUSCO_DB.nf'\n",
      "'../subworkflows/PLASMID_PREDICTION.nf'\n",
      "modules used in: subworkflows/PLASMID_PREDICTION.nf'\n",
      "'../modules/PLASMER.nf'\n",
      "'../modules/PLASMER_SORT.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../modules/VIRALVERIFY.nf'\n",
      "'../modules/PLASMER_DB.nf'\n",
      "'../modules/PFAM_DB.nf'\n",
      "'../modules/VIRSORTER2.nf'\n",
      "'../modules/VIRSORTER2_DB.nf'\n",
      "'../modules/MEF.nf'\n",
      "'../subworkflows/IDENTIFICATION.nf'\n",
      "modules used in: subworkflows/IDENTIFICATION.nf'\n",
      "'../modules/GTDBTK_DB.nf'\n",
      "'../modules/GTDB_TK.nf'\n",
      "'../subworkflows/MULTI_AMR.nf'\n",
      "modules used in: subworkflows/MULTI_AMR.nf'\n",
      "'../modules/RESFINDER_DB.nf'\n",
      "'../modules/AMRFINDER_DB.nf'\n",
      "'../modules/CARD_DB.nf'\n",
      "'../modules/AMRFINDER.nf'\n",
      "'../modules/CARD_CONTIG.nf'\n",
      "'../modules/RESFINDER.nf'\n",
      "'../modules/AMRFINDER_PARSE.nf'\n",
      "'../modules/CARD_PARSE.nf'\n",
      "'../modules/RESFINDER_PARSE.nf'\n",
      "Subworkflows in: SHORT_READ_METAGENOMIC.nf\n",
      "'../subworkflows/READ_QC_SR.nf'\n",
      "modules used in: subworkflows/READ_QC_SR.nf'\n",
      "'../modules/FASTQC.nf'\n",
      "'../modules/FASTQC.nf'\n",
      "'../modules/FASTP.nf'\n",
      "'../subworkflows/PLASMID_PREDICTION.nf'\n",
      "modules used in: subworkflows/PLASMID_PREDICTION.nf'\n",
      "'../modules/PLASMER.nf'\n",
      "'../modules/PLASMER_SORT.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../modules/VIRALVERIFY.nf'\n",
      "'../modules/PLASMER_DB.nf'\n",
      "'../modules/PFAM_DB.nf'\n",
      "'../modules/VIRSORTER2.nf'\n",
      "'../modules/VIRSORTER2_DB.nf'\n",
      "'../modules/MEF.nf'\n",
      "'../subworkflows/HOST_REMOVAL_SHORT_READ.nf'\n",
      "modules used in: subworkflows/HOST_REMOVAL_SHORT_READ.nf'\n",
      "'../modules/BOWTIE2.nf'\n",
      "'../modules/SAMTOOLS_EXTRACT_UNMAPPED.nf'\n",
      "'../modules/FASTQC.nf'\n",
      "'../modules/SAMTOOLS_STATS.nf'\n",
      "'../subworkflows/SHORT_READ_META_ASSEMBLY.nf'\n",
      "modules used in: subworkflows/SHORT_READ_META_ASSEMBLY.nf'\n",
      "'../modules/SPADES_METAGENOME.nf'\n",
      "'../modules/QUAST.nf'\n",
      "'../subworkflows/METAGENOMIC_COMMUNITY_ANALYSIS_SR.nf'\n",
      "modules used in: subworkflows/METAGENOMIC_COMMUNITY_ANALYSIS_SR.nf'\n",
      "'../modules/KRAKEN2_DB_PLUSPF.nf'\n",
      "'../modules/KRAKEN2_PLUSPF_PE.nf'\n",
      "'../modules/KRAKEN_REPORT2KRONA.nf'\n",
      "'../modules/KRAKEN_BRACKEN.nf'\n",
      "'../modules/KRAKEN_ALPHADIV.nf'\n",
      "'../modules/KRAKEN_BETADIV.nf'\n",
      "'../modules/KRAKEN_REPORT2MPA.nf'\n",
      "'../subworkflows/CARD_READS_ONLY.nf'\n",
      "modules used in: subworkflows/CARD_READS_ONLY.nf'\n",
      "'../modules/CARD_DB.nf'\n",
      "'../modules/CARD_READS.nf'\n",
      "'../subworkflows/MULTI_AMR.nf'\n",
      "modules used in: subworkflows/MULTI_AMR.nf'\n",
      "'../modules/RESFINDER_DB.nf'\n",
      "'../modules/AMRFINDER_DB.nf'\n",
      "'../modules/CARD_DB.nf'\n",
      "'../modules/AMRFINDER.nf'\n",
      "'../modules/CARD_CONTIG.nf'\n",
      "'../modules/RESFINDER.nf'\n",
      "'../modules/AMRFINDER_PARSE.nf'\n",
      "'../modules/CARD_PARSE.nf'\n",
      "'../modules/RESFINDER_PARSE.nf'\n",
      "Subworkflows in: SHORT_READ_METAGENOMIC_RNA.nf\n",
      "'../subworkflows/READ_QC_SR.nf'\n",
      "modules used in: subworkflows/READ_QC_SR.nf'\n",
      "'../modules/FASTQC.nf'\n",
      "'../modules/FASTQC.nf'\n",
      "'../modules/FASTP.nf'\n",
      "'../subworkflows/METAGENOMIC_COMMUNITY_ANALYSIS_SR.nf'\n",
      "modules used in: subworkflows/METAGENOMIC_COMMUNITY_ANALYSIS_SR.nf'\n",
      "'../modules/KRAKEN2_DB_PLUSPF.nf'\n",
      "'../modules/KRAKEN2_PLUSPF_PE.nf'\n",
      "'../modules/KRAKEN_REPORT2KRONA.nf'\n",
      "'../modules/KRAKEN_BRACKEN.nf'\n",
      "'../modules/KRAKEN_ALPHADIV.nf'\n",
      "'../modules/KRAKEN_BETADIV.nf'\n",
      "'../modules/KRAKEN_REPORT2MPA.nf'\n",
      "'../subworkflows/CARD_READS_ONLY.nf'\n",
      "modules used in: subworkflows/CARD_READS_ONLY.nf'\n",
      "'../modules/CARD_DB.nf'\n",
      "'../modules/CARD_READS.nf'\n",
      "---------------------------------------------------------------\n",
      "Directory structure created (or already exists).\n",
      "There were no unused files in subworkflows so no files were moved\n",
      "There were no unused files in modules so no files were moved\n"
     ]
    }
   ],
   "source": [
    "whole_process(pipeline_dir, workflow_dir, subworkflow_dir, module_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9eae4e38-d1e2-450b-991c-41236109b610",
   "metadata": {},
   "source": [
    "## Geting Containers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "466541bd-b384-4e5e-8cb7-c6447f02157a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "92218e12-b035-48f8-8d73-184b89c9614a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Args\n",
    "pipeline_dir = 'BALROG-MON'\n",
    "module_dir = 'modules'\n",
    "output_csv = 'BALROG-MON/test_contaienrs.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "9501370f-0a1d-4d21-8961-20aab5587d8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nf_files(folder):\n",
    "    nf_files = []\n",
    "    for entry in os.listdir(folder):\n",
    "        if os.path.isfile(os.path.join(folder, entry)) and entry.endswith('.nf'):\n",
    "            nf_files.append(os.path.join(folder, entry))\n",
    "        if os.path.isdir(os.path.join(folder, entry)):\n",
    "            new_files = get_nf_files(os.path.join(folder, entry))\n",
    "            for file in new_files:\n",
    "                nf_files.append(file)\n",
    "    return nf_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "f149a6ea-623d-4ced-89d4-96620efd0859",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_contaienrs(pipeline_dir, module_dir, output_csv):\n",
    "    \n",
    "    nf_files = get_nf_files(pipeline_dir + '/' + module_dir)\n",
    "    nf_files\n",
    "\n",
    "    paths = []\n",
    "    conts = []\n",
    "    \n",
    "    for path in nf_files:\n",
    "        with open(path) as file:\n",
    "            text = file.read().split('\\n')\n",
    "            for line in text:\n",
    "                words = line.split()\n",
    "                if 'container' in words:\n",
    "                    paths.append(path)\n",
    "                    conts.append(words[1])\n",
    "    \n",
    "    conts_df = pd.DataFrame()\n",
    "    conts_df['path'] = paths\n",
    "    conts_df['cont'] = conts\n",
    "    \n",
    "    conts_df.to_csv(output_csv, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "d3ef566d-986e-446c-91d1-9efce7bec35b",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_contaienrs(pipeline_dir, module_dir, output_csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fb12ce8-141f-4995-9505-5e3631f4a29d",
   "metadata": {},
   "source": [
    "## Getting Param Usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "51fb85a3-6524-41f3-899d-e48f5119f0ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "a21824f2-b6b8-460e-a9e9-4978d17a0a47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Args\n",
    "pipeline_dir = 'BALROG-MON'\n",
    "nextflow_configs = ['BALROG-MON/nextflow.config']\n",
    "output = 'BALROG-MON/test_params.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "151c26fc-6659-4e07-8bb7-a5487aadd04d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nf_files(folder):\n",
    "    nf_files = []\n",
    "    for entry in os.listdir(folder):\n",
    "        if os.path.isfile(os.path.join(folder, entry)) and entry.endswith('.nf'):\n",
    "            nf_files.append(os.path.join(folder, entry))\n",
    "        if os.path.isdir(os.path.join(folder, entry)):\n",
    "            new_files = get_nf_files(os.path.join(folder, entry))\n",
    "            for file in new_files:\n",
    "                nf_files.append(file)\n",
    "    return nf_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "7b3a4dbf-c215-4f67-81d7-8b47cc0d9c5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nf_files = get_nf_files(pipeline_dir)\n",
    "os.path.join(pipeline_dir, nextflow_config) in nf_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "bcdea511-e935-4749-8187-7cc21611fa6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def parse_param(line):\n",
    "    # Regular expression to match params.<variable> = 'value' or params.<variable> = \"value\"\n",
    "    match = re.search(r'params\\.(\\w+)\\s*=\\s*[\"\\']([^\"\\']*)[\"\\']', line)\n",
    "    \n",
    "    if match:\n",
    "        param_name = match.group(1)  # The parameter name (e.g., project_name)\n",
    "        param_value = match.group(2)  # The value inside quotes (e.g., BALROG_MON_RUN_1)\n",
    "        return [param_name, param_value]\n",
    "    else:\n",
    "        return None, None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "242607d9-f050-4a31-9b7a-224140171de6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def config_params(config_list):\n",
    "    params = []\n",
    "\n",
    "    for path in config_list:\n",
    "        with open(path) as file:\n",
    "            text = file.read().split('\\n')\n",
    "            line_number = 1\n",
    "            for line in text:\n",
    "                if 'params.' in line:\n",
    "                    params = parse_param(line)\n",
    "                    print(str(params) + '-----' + str(line_number))\n",
    "                    line_number+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "5cc3b5fc-2f67-4bca-bd5e-ed600420700a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['max_memory', '100.GB']-----1\n",
      "(None, None)-----2\n",
      "['max_time', '24.h']-----3\n",
      "['sample_sheet', '/bulk/ksilver/testing_data/balrong_ont/samplesheet_subset.csv']-----4\n",
      "['project_name', 'BALROG_MON_RUN_1']-----5\n",
      "['publish_dir_mode', 'copy']-----6\n",
      "['workflow_opt', 'ont_meta_rna']-----7\n",
      "['ont_min_seq_len', '500']-----8\n",
      "['ont_average_read_min_qscore', '20']-----9\n",
      "['sequencing_adapter_type', 'illumina']-----10\n",
      "(None, None)-----11\n",
      "(None, None)-----12\n",
      "['fastp_q', '20']-----13\n",
      "['busco_lineage', 'bacteria_odb10']-----14\n",
      "['plasmer_min_len', '500']-----15\n",
      "['plasmer_max_len', '500000']-----16\n",
      "['database_dir', '']-----17\n",
      "(None, None)-----18\n",
      "(None, None)-----19\n",
      "(None, None)-----20\n",
      "(None, None)-----21\n",
      "(None, None)-----22\n",
      "(None, None)-----23\n",
      "(None, None)-----24\n",
      "(None, None)-----25\n",
      "(None, None)-----26\n",
      "(None, None)-----27\n",
      "(None, None)-----28\n",
      "(None, None)-----29\n",
      "(None, None)-----30\n",
      "(None, None)-----31\n",
      "(None, None)-----32\n",
      "(None, None)-----33\n",
      "(None, None)-----34\n",
      "(None, None)-----35\n",
      "(None, None)-----36\n",
      "(None, None)-----37\n",
      "(None, None)-----38\n",
      "(None, None)-----39\n",
      "(None, None)-----40\n",
      "['fastp_adap1', 'ATGTCGGAAGGTGTGCAGGCTACCGCTTGTCAACT']-----41\n",
      "['fastp_adap2', 'ATGTCGGAAGGTGTCTGGTGAGCCAATCCAGCACG']-----42\n",
      "(None, None)-----43\n",
      "(None, None)-----44\n",
      "(None, None)-----45\n",
      "(None, None)-----46\n",
      "(None, None)-----47\n",
      "(None, None)-----48\n",
      "(None, None)-----49\n",
      "(None, None)-----50\n",
      "(None, None)-----51\n",
      "(None, None)-----52\n",
      "(None, None)-----53\n"
     ]
    }
   ],
   "source": [
    "config_params(nextflow_configs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83370c7a-6e47-44bb-8d2d-bcdb129bd91c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
